8_bit_pic.py

```diff
@@ -11,102 +37,352 @@
 logger = logging.getLogger(__name__)
 
 WORD_SIZE = 8
+FLASH_SIZE = 2 * 1024 * 1024
+RAM_SIZE = 4 * 1024
 
 
 class InstructionCodes(vm_types.GenericInstructionSet):
-    NOP = 0x00
+    NOP = 0x0000
+    MOVWF = 0x0080
+    BCF = 0x1000
+    BSF = 0x1400
+    GOTO = 0x2800
+    MOVLW = 0x3000
 
 
-TEST_PROG = """
-; Will @zayfod (Kaloyan) implement PIC 8-bit ?! 
-; *eyes*
+class LabelToken(vm_types.AssemblerToken):
+    """Label token."""
+    def __init__(self, label: str) -> None:
+        super().__init__()
+        self.label = label
+
+    def __repr__(self) -> str:
+        return f"Label('{self.label}')"
+
+    def encode(self, assembler: vm_types.GenericAssembler) -> bytes:
+        assembler.symbol_table["map"][self.label] = len(
+            assembler.byte_code
+        )
+        return b""
+
+
+class ByteOpToken(vm_types.AssemblerToken):
+    """Byte-oriented file register operation token."""
+    def __init__(self, inst: str, op: int, f: str, d: str) -> None:
+        super().__init__()
+        self.inst = str(inst)
+        self.op = int(op)
+        self.f = str(f)
+        self.d = str(d)
+
+    def __repr__(self) -> str:
+        return f"{self.inst}(f={self.f}, d={self.d})"
+
+    def encode(self, assembler: vm_types.GenericAssembler) -> bytes:
+        f = resolve_int_value(assembler, self.f, 0x7f)
+        d = resolve_int_value(assembler, self.d, 0x1)
+        code = self.op | (d << 7) | f
+        return code.to_bytes(2, 'big')
```

Comment:
I struggled a bit with the `InstructionMeta` approach as I was looking for a more readable representation of parsed tokens and ended up with this.